test/common/websocket/codec_test.cc

```diff
@@ -0,0 +1,824 @@
+  text_payload_second_frame.resize(2);
+  (*(frames.value()[1].payload_)).copyOut(0, 2, text_payload_second_frame.data());
+  EXPECT_EQ("lo", text_payload_second_frame);
+}
+
+// A fragmented and unfragmented text message
+TEST(WebSocketCodecTest, decodeFragmentedAndUnfragmentedTextMessages) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {0x01, 0x03, 0x48, 0x65, 0x6c});
+  Buffer::addSeq(buffer, {0x80, 0x02, 0x6c, 0x6f});
+  Buffer::addSeq(buffer, {0x81, 0x05, 0x48, 0x65, 0x6c, 0x6c, 0x6f});
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(3, frames.value().size());
+
+  EXPECT_EQ(false, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(3, frames.value()[0].payload_length_);
+
+  std::string text_payload_first_frame;
+  text_payload_first_frame.resize(3);
+  (*(frames.value()[0].payload_)).copyOut(0, 3, text_payload_first_frame.data());
+  EXPECT_EQ("Hel", text_payload_first_frame);
+
+  EXPECT_EQ(false, frames.value()[1].masking_key_.has_value());
+  EXPECT_EQ(2, frames.value()[1].payload_length_);
+
+  std::string text_payload_second_frame;
+  text_payload_second_frame.resize(2);
+  (*(frames.value()[1].payload_)).copyOut(0, 2, text_payload_second_frame.data());
+  EXPECT_EQ("lo", text_payload_second_frame);
+
+  EXPECT_EQ(false, frames.value()[2].masking_key_.has_value());
+  EXPECT_EQ(5, frames.value()[2].payload_length_);
+
+  std::string text_payload_third_frame;
+  text_payload_third_frame.resize(5);
+  (*(frames.value()[2].payload_)).copyOut(0, 5, text_payload_third_frame.data());
+  EXPECT_EQ("Hello", text_payload_third_frame);
+}
+
+// A single-frame masked text message ("Hello")
+// 0x81 0x85 0x37 0xfa 0x21 0x3d 0x7f 0x9f 0x4d 0x51 0x58
+//           M1   M2   M3   M4   B1   B2   B3   B4   B5
+// fin - 1
+// opcode - 0x1 - text data
+// mask bit - 1
+// length - 5 bytes 0000101
+//
+// H - M1 XOR B1   01001000 = 00110111 xor 01111111
+// e - M2 XOR B2
+// l - M3 XOR B3
+// l - M4 XOR B4
+// o - M1 XOR B5
+//
+// j = i MOD 4
+// O(i) = M(j) XOR B(i)
+TEST(WebSocketCodecTest, decodeSingleMaskedFrame) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {0x81, 0x85, 0x37, 0xfa, 0x21, 0x3d, 0x7f, 0x9f, 0x4d, 0x51, 0x58});
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(true, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(5, frames.value()[0].payload_length_);
+  EXPECT_EQ(0x37fa213d, frames.value()[0].masking_key_.value());
+
+  std::string text_payload;
+  text_payload.resize(5);
+  (*(frames.value()[0].payload_)).copyOut(0, 5, text_payload.data());
+  // Unmasking the text payload
+  for (size_t i = 0; i < text_payload.size(); ++i) {
+    text_payload[i] ^= (frames.value()[0].masking_key_.value() >> (8 * (3 - i % 4))) & 0xff;
+  }
+  EXPECT_EQ("Hello", text_payload);
```

Comment:
This block of logic is not actually testing the Decoder. It's testing that the input provided to the Decoder is able to be unmasked. But that's just testing the test, not testing the Decoder. All the decoder knows how to do is extract the masking key correctly (which is checked on line 148) and that it extracts the payload correctly. So I'd just do: `EXPECT_EQ("\7f\9f\4d\51\58", ...)` or some such.